File: 0792-binary-search/0792-binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        """
        Time: O(log n)
        Space: O(1)

        The time complexity is O(log n) since half the search space is being eliminated after 
        each iteration. 
        
        [-3, 0, 2, 6, 9]
        """
        L = 0          # inclusive
        R = len(nums)  # exclusive
        while L < R:
            mid = (L + R) // 2
            if nums[mid] > target:  # we're too big
                R = mid             # make the nums smaller
            elif nums[mid] < target:  # we're too small
                L = mid + 1           # make the nums bigger
            else:
                return mid
        return -1

File: 0792-binary-search/0792-binary-search.java
class Solution {
    public int search(int[] nums, int target) {
        int L = 0;
        int R = nums.length;

        int mid;
        while (L < R) {
            mid = (L + R) / 2;
            if (nums[mid] < target) {       // we are too small
                L = mid + 1;                // make nums bigger
            }
            else if (nums[mid] > target) {  // we are too big
                R = mid;                    // make nums smaller; R is exclusive
            }
            else {
                return mid;
            }
        }
        return -1;
    }
}